src/message.rs

```diff
@@ -257,6 +259,296 @@ impl IntoMidiMessage for ChannelVoice {
     }
 }
 
+/// Flex data messages: real time messages with limited variability of size.
+#[derive(Copy, Clone, Hash, Debug, Eq, PartialEq)]
+pub enum Flex {
+    /// Sets musical tempo by declaring the number of 10 nanosecond units per quarter note.
+    SetTempo(u32),
+    /// Declares and sets a Time Signature for subsequent bars.
+    SetTimeSignature {
+        /// A value from 1 to 256, supporting up to 256 beats in a bar.
+        numerator: u8,
+        /// A value in negative power of 2 (2 represents a quarter note, 3 represents an eighth note, etc.) If the value is set to zero, there is a non-standard denominator.
+        denominator: u8,
+        /// Expresses the number of 1/32 notes in 24 MIDI Clocks.
+        number_of_32n: u8,
+    },
+    /// Sets metronome functions.
+    SetMetronome {
+        /// Number of MIDI Clocks per Primary Click.
+        /// 1 tick is 1/96
+        /// 24 ticks is 1/4
+        clocks_per_click: u8,
+        /// Sets the number of Primary Clicks between each accent.
+        bar_accents: [u8; 3],
+        /// Sets number of clicks in between Primary Clicks.
+        subdivision_clicks: [u8; 2],
+    },
+    /// Sets the Key Signature for up to 7 sharps or up to 7 flats
+    SetKeySignature {
+        address: FlexAddress,
+        /// The number of sharps(+) or flats(-).
+        sharps_or_flats: i8,
+        /// The tonic note (A = 1, G = 7, others are non-standard)
+        tonic_note: u8,
+    },
+    /// Declares the name of a chord
+    SetChordName(FlexAddress, [u8; 12]),
+    /// Contains text encoded in UTF-8 format
+    TextMessageCommonFormat {
+        /// determines the address destination of each UMP: group, channel, or <reserved>
+        address: FlexAddress,
+        /// determines the role of each UMP in a Flex Data Message
+        format: FlexFormat,
+        /// provides up to 256 (text) message classifications
+        status: FlexStatus,
+        data: [u8; 12],
+    },
+}
+
+/// Determines the destination of each UMP.
+#[derive(Copy, Clone, Hash, Debug, Eq, PartialEq)]
+pub enum FlexAddress {
+    Channel(u8),
+    Group(u8),
+    _Reserved1(u8),
+    _Reserved2(u8),
+}
+
+impl From<FlexAddress> for u8 {
+    fn from(value: FlexAddress) -> Self {
+        match value {
+            FlexAddress::Channel(channel) => 0 << 4 | channel,
+            FlexAddress::Group(group) => 1 << 4 | group,
+            FlexAddress::_Reserved1(cccc) => 2 << 4 | cccc,
+            FlexAddress::_Reserved2(cccc) => 3 << 4 | cccc,
+        }
+    }
+}
+
+impl From<u8> for FlexAddress {
+    fn from(value: u8) -> Self {
```

Comment:
A refactor that I plan on doing is replacing any `From` with `TryFrom`, as this code is not sound for malformed streams.